Google Cloud Storage physical backend

tallpauley opened the pull request:
Implementing #1305: Google Cloud Storage for Storage Backend

### Testing
```
export GOOGLE_APPLICATION_CREDENTIALS=myserviceaccount.json
# project id is only required for testing since we need to know where to create the test bucket
export GOOGLE_PROJECT_ID=gcp-project-to-test
make testacc TEST=./physical TESTARGS='-run GCPBackend'
```

### Notes:
* google storage has [eventual consistency](https://cloud.google.com/storage/docs/consistency) for list-after-write, and list-after-delete sequences:
  *  so the acceptance tests don't pass 100% of the time
  * we shouldn't expect data stores with eventually consistent list operations to always pass the acceptance test (which does immediate list-after-write and list-after-delete many times)
* I did `govendor fetch cloud.google.com/go/storage` which brought in a bunch of files we don't need from **google.golang.org/appengine**, so I removed the files manually and the related changes in **vendor.json**
* I was unable to get [credentials serialized into an environment variable](https://github.com/hashicorp/vault/issues/1305#issuecomment-258196642) working, so this could be implemented as a feature later
* I based my backend off of **s3.go**, so thanks to the authors of that backend

jefferai commented:
> google storage has eventual consistency for list-after-write, and list-after-delete sequences:
so the acceptance tests don't pass 100% of the time

I think then that the acceptance tests should be forked for this backend, or modified to do a sleep and try again a few times. Listing helps ensure that expected values are seen, and eventually we need lists to show the correct results even if it means waiting a bit.

> I did govendor fetch cloud.google.com/go/storage which brought in a bunch of files we don't need from google.golang.org/appengine, so I removed the files manually and the related changes in vendor.json

This needs to be fixed; it's not sustainable to manually make changes after every update

> I was unable to get credentials serialized into an environment variable working, so this could be implemented as a feature later

Isn't this the recommended approach by Google? If so we should support it. If not, what issues did you run into?

tallpauley commented:
I amended my commit with the following modifications:

> google storage has eventual consistency for list-after-write, and list-after-delete sequences:
so the acceptance tests don't pass 100% of the time

>> I think then that the acceptance tests should be forked for this backend, or modified to do a sleep and try again a few times. Listing helps ensure that expected values are seen, and eventually we need lists to show the correct results even if it means waiting a bit.

I duplicated the functions `testBackend()` and  `testBackend_ListPrefix()` to `testEventuallyConsistentBackend()` and `testEventuallyConsistentBackend_ListPrefix()` which take a struct:
```
type delays struct {
	beforeGet  time.Duration
	beforeList time.Duration
}
```
This way the test can be used for other physical backends with eventually consistency. I did a 5 second delay before listing for GCP since at 2 seconds I still occasionally failed the test. I duplicated the test functions and modified because I figured that would be more intuitive than parameterizing the standard tests. I also avoided retries to not complicate the test logic.

> I did govendor fetch cloud.google.com/go/storage which brought in a bunch of files we don't need from google.golang.org/appengine, so I removed the files manually and the related changes in vendor.json

>> This needs to be fixed; it's not sustainable to manually make changes after every update

Fixed this, just let `govendor fetch cloud.google.com/go/storage` do its thing. 

> I was unable to get credentials serialized into an environment variable working, so this could be implemented as a feature later

>> Isn't this the recommended approach by Google? If so we should support it. If not, what issues did you run into?

The [Google recommended approach](https://developers.google.com/identity/protocols/application-default-credentials) is environment variable ```GOOGLE_APPLICATION_CREDENTIALS ``` which points to the JSON file. The serialized JSON in environment variable was just a nice-to-have. I used [this code](https://github.com/hashicorp/vault/issues/1305#issuecomment-258196642) to try to load the same json file from environment variable, but it gave an error about missing the scope. Once again though, this is a feature that could be added later, and is not Google's recommended approach

kelseyhightower commented:
@tallpauley I would suggest this backend be renamed to gcs (Google Cloud Storage) since gcp (Google Cloud Platform) is a generic reference to the entire Google Cloud offering.

tallpauley commented:
@kelseyhightower Sure, went ahead and changed it to `gcs` everywhere. The backend name in configuration is also now `gcs` instead of `google`.

jefferai commented on physical/gcs.go:
```diff
@@ -0,0 +1,182 @@
+	if bucketName == "" {
+		bucketName = conf["bucket"]
+		if bucketName == "" {
+			return nil, fmt.Errorf("'bucket' must be set")
```
Maybe put, in the error, something like `the GOOGLE_STORAGE_BUCKET env var or 'bucket' configuration parameter must be set` to make it clear both can work.

jefferai commented on physical/gcs.go:
```diff
@@ -0,0 +1,182 @@
+	writer := bucket.Object(entry.Key).NewWriter(context.Background())
+
+	_, err := writer.Write(entry.Value)
+	if cerr := writer.Close(); cerr != nil && err == nil {
```
I'm not in favor of returning an error on close. The reason is that doing so will prevent various other parts of Vault (such as the read cache) from acknowledging that the operation was successful, even though the write itself was (unless the write may not be successful with GCS until Close is called). This can then cause the cache and underlying data store to think that the contained values are different.

I would instead propose simply logging errors to the logger if you see them on close.

jefferai commented on physical/gcs.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+// Put is used to insert or update an entry
+func (s *GCSBackend) Put(entry *Entry) error {
```
S3 doesn't have this (because nobody has stepped in to do it) but take a look at the Consul backend's permitPool and max_parallel parameter. We often see people wanting to tune this even if the native client doesn't let you. I think for a new backend this should be in there from the beginning. (This affects multiple operations, just putting the comment on Put.)

jefferai commented on physical/gcs.go:
```diff
@@ -0,0 +1,182 @@
+
+// Put is used to insert or update an entry
+func (s *GCSBackend) Put(entry *Entry) error {
+	defer metrics.MeasureSince([]string{"google storage", "put"}, time.Now())
```
Everywhere else uses `_` (or, more legacy, `-`) to separate words in metrics...please use `_` to keep things more consistent.

tallpauley commented on physical/gcs.go:
```diff
@@ -0,0 +1,205 @@
+	if bucketName == "" {
+		bucketName = conf["bucket"]
+		if bucketName == "" {
+			return nil, fmt.Errorf("env var GOOGLE_STORAGE_BUCKET or configuration parameter 'bucket' must be set")
```
request 1: make config error messages more explicit

tallpauley commented on physical/gcs.go:
```diff
@@ -0,0 +1,205 @@
+		return nil, fmt.Errorf("unable to access bucket '%s': '%v'", bucketName, err)
+	}
+
+	maxParStr, ok := conf["max_parallel"]
```
request 2: use `"max_parallel"` and `PermitPool` for all requests (`Put`, `Get`, `Delete`, `Write`)

tallpauley commented on physical/gcs.go:
```diff
@@ -0,0 +1,205 @@
+func (g *GCSBackend) Put(entry *Entry) error {
+	defer metrics.MeasureSince([]string{"gcs", "put"}, time.Now())
+
+	bucket := g.client.Bucket(g.bucketName)
```
use `g` instead of `s` for references to GCSBackend object

tallpauley commented on physical/gcs.go:
```diff
@@ -0,0 +1,205 @@
+	g.permitPool.Acquire()
+	defer g.permitPool.Release()
+
+	defer writer.Close()
```
request 4: don't return an error on `Close()`. I just modeled it after `file.go` where we don't check for errors on `Close()`

tallpauley commented on physical/gcs.go:
```diff
@@ -0,0 +1,205 @@
+
+// Put is used to insert or update an entry
+func (g *GCSBackend) Put(entry *Entry) error {
+	defer metrics.MeasureSince([]string{"gcs", "put"}, time.Now())
```
request 3: remove space from "google storage". I did "gcs" instead.

tallpauley commented on website/source/docs/config/index.html.md:
```diff
@@ -611,6 +614,17 @@ will cause Vault to attempt to retrieve credentials from the metadata service.
+
+  * `credentials_file` - (required) The path to a GCP [service account](https://cloud.google.com/compute/docs/access/service-accounts) private key file in [JSON format](https://cloud.google.com/storage/docs/authentication#generating-a-private-key). It must be provided, but it can also be sourced from the `GOOGLE_APPLICATION_CREDENTIALS` environment variable.
+
+  * `max_parallel` (optional) - The maximum number of concurrent requests to Google Cloud Storage.
```
add `max_parallel` to docs

jefferai commented:
Thanks for the updates!
